src/v/cloud_io/basic_cache_service_api.h

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2021 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+#include "base/units.h"
+
+#include <seastar/core/file.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/io_priority_class.hh>
+#include <seastar/core/iostream.hh>
+#include <seastar/core/lowres_clock.hh>
+
+#include <filesystem>
+#include <iterator>
+#include <optional>
+#include <set>
+#include <string_view>
+
+namespace cloud_io {
+
+static constexpr size_t default_write_buffer_size = 128_KiB;
+static constexpr unsigned default_write_behind = 10;
+static constexpr size_t default_read_buffer_size = 128_KiB;
+static constexpr unsigned default_read_ahead = 10;
+
+struct [[nodiscard]] cache_item {
+    ss::file body;
+    size_t size;
+};
+
+struct [[nodiscard]] cache_item_str {
+    ss::input_stream<char> body;
+    size_t size;
+};
+
+enum class [[nodiscard]] cache_element_status {
+    available,
+    not_available,
+    in_progress,
+};
+
+std::ostream& operator<<(std::ostream& o, cache_element_status);
+
+template<class Clock>
+class basic_cache_service_api;
+
+/// RAII guard for bytes reserved in the cache: constructed prior to a call
+/// to cache::put, and may be destroyed afterwards.
+template<class Clock>
+class basic_space_reservation_guard {
+public:
+    basic_space_reservation_guard(
+      basic_cache_service_api<Clock>& cache,
+      uint64_t bytes,
+      size_t objects) noexcept
+      : _cache(cache)
+      , _bytes(bytes)
+      , _objects(objects) {}
+
+    basic_space_reservation_guard(const basic_space_reservation_guard&)
+      = delete;
+    basic_space_reservation_guard() = delete;
+    basic_space_reservation_guard(basic_space_reservation_guard&& rhs) noexcept
+      : _cache(rhs._cache)
+      , _bytes(rhs._bytes)
+      , _objects(rhs._objects) {
+        rhs._bytes = 0;
+        rhs._objects = 0;
+    }
+
+    ~basic_space_reservation_guard();
+
+    /// After completing the write operation that this space reservation
+    /// protected, indicate how many bytes were really written: this is used to
+    /// atomically update cache usage stats to free the reservation and update
+    /// the bytes used stats together.
+    ///
+    /// May only be called once per reservation.
+    void wrote_data(uint64_t, size_t);
```

Comment:
Not for this PR but can't we use value semantics for the reservation and make this `void wrote_data(...) &&`? For some reason we are passing reservation around as reference.

Reply:
Would the assembly look much worse?

Reply:
the important thing in this pr is that we are preserving the original semantics. are we?